sdks/python/apache_beam/runners/common.py

```diff
@@ -437,13 +439,15 @@ def invoke_start_bundle(self):
     # type: () -> None
     """Invokes the DoFn.start_bundle() method.
     """
+    assert self.output_processor is not None
```

Comment:
These asserts should be fast, but have you verified this doesn't impact performance (given this is called for every element of every transform). Or is there another way to declare this for critical parts of the code. (In particular, isinstance can be slow. It's not common convention to disable asserts in Python.)

Reply:
I have not done performance testing on this.  Is there a way that we can invoke the perf suite on Jenkins from github?

The underlying issue in this case is that it's possible to instantiate a `DoFnInvoker` without an `output_processor` and that's considered ok as long as you don't call any of the methods that use the `output_processor`.   If you do, then it would raise an exception.  The asserts obviously also raise an exception, but it serves as a way to communicate to mypy that you're aware of it, and it can adjust its type analysis within that scope.

**Solutions**

Easy:  Judiciously add `type: ignore` comments.  I say "judiciously" because ignoring an error does not change the type analysis, so similar errors can pop up nearby in the same scope.  In this particular case the methods are brief, so a `type: ignore` comment would suffice.

Not Easy:  Rework the code so there's a subclass of `DoFnInvoker` that always has a non-optional `output_processor`, and only this class possesses these "safe if you're careful" methods, which would, under that design, always be safe.

The easy solution is fine for this particular case (we're aware there could be an error, but we accept it), but it's not a general solution to this problem.  Choosing the right solution for each case takes some consideration.

Reply:
I think we can get rid of `output_processor` altogether in this code... (It's a bit ugly ever since it was introduced, this typing confirms that.) May not be the case everywhere though.

Reply:
Ok, how about I change these to `type: ignore` and you create a ticket to remove `output_processor`?  I would make the ticket but I don't think I have enough context to explain it properly.

Reply:
sgtm

Reply:
This is done.  I ended up leaving one assert in `invoke_process` because it was conditional, and by using one assert I avoided 3 ignores that cluttered up the code.  Let me know what you think.